core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java

```diff
@@ -636,21 +545,18 @@ public boolean equals(Object o) {
 
     @Override
     public int hashCode() {
-        if (frozen && usePreComputedHashCode) {
```

Comment:
can we just pre-compute the hashCode in the ctor instead? I can already see this going bad with concurrency and we might be just lucky we didn't hit that yet? Either we do in the ctor and make it final or we need a volatile read here